packages/babel-core/src/transformation/block-hoist-plugin.js

```diff
@@ -41,23 +39,29 @@ const blockHoistPlugin = {
   visitor: {
     Block: {
       exit({ node }) {
-        let hasChange = false;
-        for (let i = 0; i < node.body.length; i++) {
-          const bodyNode = node.body[i];
-          if (bodyNode?._blockHoist != null) {
-            hasChange = true;
-            break;
+        if (!node.body.find(node => node?._blockHoist != null)) return;
+        // TODO: Babel 9 (?) - remove stabilityMap once Node >= 12 is guaranteed
+        // Array sorting is not stable in earlier Node releases
+        // See: https://v8.dev/blog/array-sort for details
+        const nodeVersion =
+          typeof process !== "undefined"
+            ? process.versions.node.split(".")[0]
+            : undefined;
+        const stabilityMap =
+          nodeVersion && nodeVersion < 12
```

Comment:
If `nodeVersion` is not defined (browsers) we should still use the map because the sorting might not be stable.

```suggestion
          !nodeVersion || nodeVersion < 12
```